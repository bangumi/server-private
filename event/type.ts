export const EventOp = Object.freeze({
  Create: 'c',
  Delete: 'd',
  Update: 'u',
  Snapshot: 'r',
} as const);

export type EventOp = (typeof EventOp)[keyof typeof EventOp];

export interface Payload {
  // before: object;
  // after: object;
  source: {
    table: string;
  };
  op: EventOp;
}

export interface KafkaMessage {
  topic: string;
  key: string;
  value: string;
}
