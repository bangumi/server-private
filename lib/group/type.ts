export enum GroupTopicMode {
  All = 'all',
  Joined = 'joined',
}

export enum GroupSort {
  Trends = 'trends',
  Created = 'created',
  Updated = 'updated',
}
