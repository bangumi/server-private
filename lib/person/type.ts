export enum PersonCat {
  Character = 'crt',
  Person = 'prsn',
}
