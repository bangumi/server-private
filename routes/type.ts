import type { IncomingMessage, ServerResponse } from 'node:http';

import type { Static, TSchema } from '@sinclair/typebox';
import type {
  ContextConfigDefault,
  FastifyBaseLogger,
  FastifyInstance,
  FastifyReply,
  FastifyRequest,
  FastifySchema,
  FastifyTypeProvider,
  RawReplyDefaultExpression,
  RawRequestDefaultExpression,
  RawServerDefault,
  RouteGenericInterface,
  RouteHandlerMethod,
} from 'fastify';
import type { ResolveFastifyRequestType } from 'fastify/types/type-provider.d.ts';

// from https://github.com/fastify/fastify-type-provider-typebox/blob/v3.1.0/index.ts#L56
export interface TypeBoxTypeProvider extends FastifyTypeProvider {
  output: this['input'] extends TSchema ? Static<this['input']> : never;
}

export type App = FastifyInstance<
  RawServerDefault,
  IncomingMessage,
  ServerResponse,
  FastifyBaseLogger,
  TypeBoxTypeProvider
>;

export interface Option {
  tags?: string[];
}

export type Request<schema extends FastifySchema> = FastifyRequest<
  RouteGenericInterface,
  RawServerDefault,
  RawRequestDefaultExpression,
  schema,
  TypeBoxTypeProvider,
  ContextConfigDefault,
  FastifyBaseLogger,
  ResolveFastifyRequestType<TypeBoxTypeProvider, schema, RouteGenericInterface>
>;

export type Reply<schema extends FastifySchema> = FastifyReply<
  RawServerDefault,
  RawRequestDefaultExpression,
  RawReplyDefaultExpression,
  RouteGenericInterface,
  ContextConfigDefault,
  schema,
  TypeBoxTypeProvider
>;

export type Handler<schema extends FastifySchema> = RouteHandlerMethod<
  RawServerDefault,
  RawRequestDefaultExpression,
  RawReplyDefaultExpression,
  RouteGenericInterface,
  ContextConfigDefault,
  schema,
  TypeBoxTypeProvider
>;
