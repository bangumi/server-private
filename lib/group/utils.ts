import { db, op, schema } from '@app/drizzle';

/** Is user member of group */
export async function isMemberInGroup(userID: number, groupID: number): Promise<boolean> {
  const [d] = await db
    .select({ uid: schema.chiiGroupMembers.uid, gid: schema.chiiGroupMembers.gid })
    .from(schema.chiiGroupMembers)
    .where(
      op.and(
        op.eq(schema.chiiGroupMembers.uid, userID),
        op.eq(schema.chiiGroupMembers.gid, groupID),
      ),
    )
    .limit(1);

  return Boolean(d);
}
