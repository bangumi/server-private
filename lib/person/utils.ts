import { db, op, schema } from '@app/drizzle';

import type { PersonCat } from './type';

export async function getPersonCollect(
  cat: PersonCat,
  userID: number,
  personID: number,
): Promise<number | null> {
  const [d] = await db
    .select()
    .from(schema.chiiPersonCollects)
    .where(
      op.and(
        op.eq(schema.chiiPersonCollects.cat, cat),
        op.eq(schema.chiiPersonCollects.uid, userID),
        op.eq(schema.chiiPersonCollects.mid, personID),
      ),
    )
    .limit(1);
  if (!d) {
    return null;
  }
  return d.createdAt;
}
