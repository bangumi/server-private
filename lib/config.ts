/**
 * 从 `projectRoot/config.yaml` 和环境变量读取配置。 配置文件见 `config.example.yaml`。
 *
 * {@link schema} 定义了配置文件的 json schema
 *
 * 可以使用环境变量会覆盖对应的文件配置，如 `MYSQL_DB` 会覆盖 `mysql.db`, 具体可用的环境变量和对应的配置项也定义在 {@link schema}。
 *
 * 除此之外可以设置的环境变量：
 *
 * {@link NODE_ENV}
 *
 * 'production' | 'stage' | 'test' 在测试时会被 vitest 会设置此环境变量为 'test'， 在生产环境会被设置为 'production',
 * 在测试部属环境被会设置为 stage
 *
 * {@link HTTPS_PROXY}
 *
 * 默认为空，如果设置了的话，会作为 turnstile 的代理
 *
 * {@link REF}
 *
 * 用于 docker 镜像，不用设置。
 *
 * @packageDocumentation
 */
import * as fs from 'node:fs';
import * as path from 'node:path';
import * as process from 'node:process';
import * as url from 'node:url';

import type { Static, TSchema } from '@sinclair/typebox';
import { Kind, Type as t } from '@sinclair/typebox';
import { Value } from '@sinclair/typebox/value';
import Ajv from 'ajv';
import addFormats from 'ajv-formats';
import * as yaml from 'js-yaml';
import * as lo from 'lodash-es';

// read from env

const { HTTPS_PROXY = '', NODE_ENV, REF } = process.env;

export const production = NODE_ENV === 'production';
export const stage = NODE_ENV === 'stage';
export const developing = NODE_ENV === 'development';
export const testing = NODE_ENV === 'test';

export const projectRoot = url.fileURLToPath(new URL('..', import.meta.url));
export const pkg = JSON.parse(
  fs.readFileSync(path.resolve(projectRoot, 'package.json'), 'utf8'),
) as { version: string };

export const VERSION = developing ? 'development' : REF || pkg.version;

export const redisPrefix = `graphql-${VERSION}`;

export { HTTPS_PROXY };

/** WARNING: 所有的 key 都必需小写，而且不能包含下划线。 */
const schema = t.Object({
  nsfw_word: t.Optional(t.String({ minLength: 1 })),
  disable_words: t.Optional(t.String()),
  banned_domain: t.Optional(t.String()),

  redisUri: t.String({ default: 'redis://127.0.0.1:3306/0', env: 'REDIS_URI' }),

  image: t.Object({
    provider: t.Enum(
      {
        s3: 's3',
        FS: 'fs',
      } as const,
      { default: 'fs', env: 'CHII_IMAGE_PROVIDER' },
    ),
    imaginaryUrl: t.Optional(
      t.String({
        description: 'url to docker image running https://github.com/h2non/imaginary',
      }),
    ),
    fs: t.Object({
      path: t.String({ default: './tmp/images' }),
    }),
    s3: t.ReadonlyOptional(
      t.Object({
        endPoint: t.String({ env: 'CHII_IMAGE_S3_ENDPOINT' }),
        bucket: t.String({ default: 'chii-image', env: 'CHII_IMAGE_S3_BUCKET' }),
        port: t.Integer({ default: 9000, env: 'CHII_IMAGE_S3_PORT' }),
        useSSL: t.Boolean({ default: false, env: 'CHII_IMAGE_S3_USE_SSL' }),
        accessKey: t.String({ env: 'CHII_IMAGE_S3_ACCESS_KEY' }),
        secretKey: t.String({ env: 'CHII_IMAGE_S3_SECRET_KEY' }),
      }),
    ),
  }),

  turnstile: t.Object({
    secretKey: t.String({
      default: '1x0000000000000000000000000000000AA',
      env: 'TURNSTILE_SECRET_KEY',
    }),
    siteKey: t.String({ default: '1x00000000000000000000AA', env: 'TURNSTILE_SITE_KEY' }),
  }),

  mysql: t.Object({
    db: t.String({ default: 'bangumi', env: 'MYSQL_DB' }),
    host: t.String({ default: '127.0.0.1', env: 'MYSQL_HOST' }),
    port: t.Integer({ default: 3306, env: 'MYSQL_PORT' }),
    user: t.String({ default: 'user', env: 'MYSQL_USER' }),
    password: t.String({ default: 'password', env: 'MYSQL_PASS' }),
  }),
});

// read config file

export default readConfig();

function readConfig(): Static<typeof schema> {
  const configFilePath = path.resolve(projectRoot, 'config.yaml');

  let configFileContent = '{}';
  if (fs.existsSync(configFilePath)) {
    configFileContent = fs.readFileSync(configFilePath, 'utf8');
  }

  const config = lo.assign(Value.Create(schema), yaml.load(configFileContent));

  function readFromEnv(keyPath: string[], o: TSchema) {
    if (o[Kind] === 'Object') {
      for (const [key, value] of Object.entries(o.properties as Record<string, TSchema>)) {
        readFromEnv([...keyPath, key], value);
      }

      return;
    }

    const envKey = o.env as string | undefined;
    if (envKey) {
      const v = process.env[envKey];
      if (v) {
        lo.set(config, keyPath, process.env[envKey]);
      }
    }
  }

  readFromEnv([], schema);

  const ajv = new Ajv({ allErrors: true, coerceTypes: true, keywords: ['env'] });
  addFormats(ajv);

  const check = ajv.compile(schema);

  const valid = check(config);

  if (!valid) {
    const errorMessage =
      check.errors
        ?.map(
          (x) => '  ' + (x.instancePath + ': ' + (x.message ?? `wrong data type ${x.schemaPath}`)),
        )
        .join('\n') ?? '';

    throw new TypeError('failed to validate config file:\n' + errorMessage);
  }

  return config;
}
