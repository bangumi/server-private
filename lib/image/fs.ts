import * as fs from 'node:fs/promises';
import * as path from 'node:path';

import config from '@app/lib/config';

const BasePath = config.image.fs.path;

export async function uploadImage(p: string, content: Buffer): Promise<void> {
  p = path.join(BasePath, p);

  await fs.mkdir(path.dirname(p), { recursive: true });
  await fs.writeFile(p, content);
}
