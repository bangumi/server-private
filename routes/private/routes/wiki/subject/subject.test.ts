import * as fs from 'node:fs/promises';
import * as path from 'node:path';

import { StatusCodes } from 'http-status-codes';
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';

import { UserGroup } from '@app/lib/auth/index.ts';
import { projectRoot } from '@app/lib/config.ts';
import * as image from '@app/lib/image/index.ts';
import type { Permission } from '@app/lib/orm/index.ts';
import { SubjectRepo } from '@app/lib/orm/index.ts';
import type { IImaginary, Info } from '@app/lib/services/imaginary.ts';
import * as Subject from '@app/lib/subject/index.ts';
import { SubjectType } from '@app/lib/subject/index.ts';
import type { ISubjectEdit, ISubjectNew } from '@app/routes/private/routes/wiki/subject/index.ts';
import { setup } from '@app/routes/private/routes/wiki/subject/index.ts';
import { createTestServer } from '@app/tests/utils.ts';
import { DateTime } from 'luxon';

async function testApp(...args: Parameters<typeof createTestServer>) {
  const app = createTestServer(...args);
  await app.register(setup);
  return app;
}

const newSubjectApp = () =>
  testApp({
    auth: {
      groupID: UserGroup.Normal,
      login: true,
      permission: { subject_edit: true },
      allowNsfw: true,
      regTime: 0,
      userID: 100,
    },
  });

describe('create subject', () => {
  test('create new subject', async () => {
    const app = await newSubjectApp();
    const res = await app.inject({
      url: '/subjects',
      method: 'post',
      payload: {
        name: 'New Subject',
        infobox: `{{Infobox
        | 话数 = 10
        }}`,
        type: SubjectType.Anime,
        platform: 0,
        metaTags: ['WEB', '3D'],
        summary: 'A brief summary of the subject',
        nsfw: false,
      } satisfies ISubjectNew,
    });

    expect(res.statusCode).toBe(200);

    expect(res.json()).toEqual({
      subjectID: expect.any(Number),
    });
  });

  test('create type', async () => {
    const app = await newSubjectApp();
    const res = await app.inject({
      url: '/subjects',
      method: 'post',
      payload: {
        name: 'New Subject',
        infobox: `{{Infobox
        | 话数 = 10
        }}`,
        type: 0,
        platform: 0,
        metaTags: ['WEB', '3D'],
        summary: 'A brief summary of the subject',
        nsfw: false,
      },
    });

    expect(res.statusCode).toBe(StatusCodes.BAD_REQUEST);

    expect(res.json()).toMatchInlineSnapshot(`
      Object {
        "code": "REQUEST_VALIDATION_ERROR",
        "error": "Bad Request",
        "message": "body/type must be equal to one of the allowed values",
        "statusCode": 400,
      }
    `);
  });

  test('invalid platform', async () => {
    const app = await newSubjectApp();
    const res = await app.inject({
      url: '/subjects',
      method: 'post',
      payload: {
        name: 'New Subject',
        infobox: `{{Infobox
        | 话数 = 10
        }}`,
        type: SubjectType.Anime,
        platform: 777777888,
        metaTags: ['WEB', '3D'],
        summary: 'A brief summary of the subject',
        nsfw: false,
      },
    });

    expect(res.statusCode).toBe(StatusCodes.BAD_REQUEST);

    expect(res.json()).toMatchInlineSnapshot(`
      Object {
        "code": "BAD_REQUEST",
        "error": "Bad Request",
        "message": "条目分类错误",
        "statusCode": 400,
      }
    `);
  });
});

describe('edit subject ', () => {
  const editSubject = vi.fn();

  beforeEach(() => {
    vi.spyOn(Subject, 'edit').mockImplementation(editSubject);
  });

  afterEach(() => {
    editSubject.mockReset();
  });

  test('should get current wiki info', async () => {
    const app = await testApp({});

    const res = await app.inject('/subjects/8');

    expect(res.json()).toMatchSnapshot();
  });

  test('should get edit history', async () => {
    const app = createTestServer({});
    await app.register(setup);

    const res = await app.inject('/subjects/8/history-summary');

    expect(res.json()).toMatchSnapshot();
  });

  test('should need authorization', async () => {
    const payload = {
      subject: {
        name: 'n',
        infobox: 'i',
        platform: 0,
        summary: 's',
        date: '0000-00-00',
        nsfw: false,
        metaTags: [],
      } satisfies ISubjectEdit,
      commitMessage: 'c',
    };
    const app = await testApp({
      auth: {
        groupID: UserGroup.Normal,
        login: true,
        permission: {},
        allowNsfw: true,
        regTime: 0,
        userID: 100,
      },
    });

    const res = await app.inject({
      url: '/subjects/1',
      method: 'put',
      payload,
    });

    expect(res.json()).toMatchInlineSnapshot(`
      Object {
        "code": "NOT_ALLOWED",
        "error": "Unauthorized",
        "message": "you don't have permission to edit subject",
        "statusCode": 401,
      }
    `);
    expect(res.statusCode).toBe(401);
  });

  test('should update subject', async () => {
    const payload = {
      subject: {
        name: 'n',
        infobox: 'i',
        platform: 0,
        metaTags: [],
        summary: 's',
        date: '0000-00-00',
        nsfw: false,
      } satisfies ISubjectEdit,
      commitMessage: 'c',
    };
    const app = await testApp({
      auth: {
        groupID: UserGroup.Normal,
        login: true,
        permission: { subject_edit: true },
        allowNsfw: true,
        regTime: 0,
        userID: 100,
      },
    });

    const res = await app.inject({
      url: '/subjects/1',
      method: 'put',
      payload,
    });

    expect(res.statusCode).toBe(200);
    expect(editSubject).toBeCalledWith({
      commitMessage: 'c',
      infobox: 'i',
      name: 'n',
      platform: 0,
      subjectID: 1,
      now: expect.any(DateTime),
      date: '0000-00-00',
      nsfw: false,
      metaTags: [],
      summary: 's',
      userID: 100,
    });
  });
});

describe('should upload image', () => {
  const uploadImageMock = vi.fn();

  vi.spyOn(image, 'uploadSubjectImage').mockImplementation(uploadImageMock);
  vi.spyOn(Subject, 'uploadCover').mockImplementation(() => Promise.resolve());

  afterEach(() => {
    uploadImageMock.mockReset();
  });

  // only allow images in ./fixtures/
  vi.mock('@app/lib/services/imaginary', async () => {
    const mod = await vi.importActual<typeof import('@app/lib/services/imaginary.ts')>(
      '@app/lib/services/imaginary',
    );

    const images = await Promise.all(
      ['webp', 'jpg'].map(async (ext) => {
        return {
          ext,
          content: await fs.readFile(path.join(projectRoot, `lib/image/fixtures/subject.${ext}`)),
        };
      }),
    );

    expect(images).toHaveLength(2);

    return {
      default: {
        async info(img: Buffer) {
          const i = images.find((x) => x.content.equals(img));
          if (i) {
            return { type: i.ext } as Info;
          }
          throw new mod.NotValidImageError();
        },

        convert(): Promise<Buffer> {
          return Promise.resolve(Buffer.from(''));
        },
      } satisfies IImaginary,
    };
  });

  test('upload subject cover', async () => {
    const app = await testApp({
      auth: {
        groupID: UserGroup.Normal,
        login: true,
        permission: { subject_edit: true },
        allowNsfw: true,
        regTime: 0,
        userID: 100,
      },
    });

    const res = await app.inject({
      url: '/subjects/184017/covers',
      method: 'post',
      payload: {
        content: Buffer.from('hello world').toString('base64'),
      },
    });

    expect(res.json()).toMatchInlineSnapshot(`
      Object {
        "code": "ERR_NOT_IMAGE",
        "error": "Bad Request",
        "message": "invalid image file",
        "statusCode": 400,
      }
    `);
    expect(res.statusCode).toBe(400);
  });

  test.each(['jpg', 'webp'])('upload subject covers in %s format', async (format) => {
    const app = await testApp({
      auth: {
        groupID: UserGroup.Normal,
        login: true,
        permission: { subject_edit: true },
        allowNsfw: true,
        regTime: 0,
        userID: 100,
      },
    });

    const raw = await fs.readFile(path.join(projectRoot, `lib/image/fixtures/subject.${format}`));

    const res = await app.inject({
      url: '/subjects/184017/covers',
      method: 'post',
      payload: {
        content: raw.toString('base64'),
      },
    });

    expect(res.statusCode).toBe(200);
    expect(uploadImageMock).toBeCalledWith(
      expect.stringMatching(/.*\.jpe?g$/),
      expect.objectContaining({}),
    );
  });
});

const lockSubjectApp = () =>
  testApp({
    auth: {
      groupID: UserGroup.Normal,
      login: true,
      permission: { subject_edit: true, subject_lock: true } as Permission,
      allowNsfw: true,
      regTime: 0,
      userID: 100,
    },
  });

describe('lock subject', () => {
  const subjectID = 12;
  test('lock subject', async () => {
    const app = await lockSubjectApp();
    const res = await app.inject({
      url: '/lock/subjects',
      method: 'post',
      payload: {
        reason: 'string',
        subjectID: subjectID,
      },
    });

    expect(res.json()).toMatchInlineSnapshot(`Object {}`);
    expect(res.statusCode).toBe(200);

    const subject = await SubjectRepo.findOneBy({ id: subjectID });
    expect(subject?.subjectBan).toBe(1);
  });

  test('unlock subject', async () => {
    const app = await lockSubjectApp();
    const res = await app.inject({
      url: '/unlock/subjects',
      method: 'post',
      payload: {
        reason: 'string',
        subjectID: subjectID,
      },
    });

    expect(res.json()).toMatchInlineSnapshot(`Object {}`);
    expect(res.statusCode).toBe(200);

    const subject = await SubjectRepo.findOneBy({ id: subjectID });
    expect(subject?.subjectBan).toBe(0);
  });
});

describe('create episodes', () => {
  const subjectID = 13;
  const newEpisodeApp = () =>
    testApp({
      auth: {
        groupID: UserGroup.Normal,
        login: true,
        permission: { ep_edit: true },
        allowNsfw: true,
        regTime: 0,
        userID: 100,
      },
    });

  test('create new episodes', async () => {
    const app = await newEpisodeApp();
    const res = await app.inject({
      url: `/subjects/${subjectID}/ep`,
      method: 'post',
      payload: {
        episodes: [
          {
            name: 'Episode 1',
            nameCN: '第一话',
            type: 0,
            ep: 1,
            duration: '24:00',
            date: '2024-01-01',
            summary: 'First episode summary',
          },
          {
            ep: 2,
          },
        ],
      },
    });

    expect(res.statusCode).toBe(200);
    expect(res.json()).toEqual({
      episodeIDs: expect.arrayContaining([expect.any(Number), expect.any(Number)]),
    });
  });

  test('should require login', async () => {
    const app = await testApp({});
    const res = await app.inject({
      url: `/subjects/${subjectID}/ep`,
      method: 'post',
      payload: {
        episodes: [{ ep: 1 }],
      },
    });

    expect(res.statusCode).toBe(401);
  });

  test('should require ep_edit permission', async () => {
    const app = await testApp({
      auth: {
        groupID: UserGroup.Normal,
        login: true,
        permission: {},
        allowNsfw: true,
        regTime: 0,
        userID: 100,
      },
    });
    const res = await app.inject({
      url: `/subjects/${subjectID}/ep`,
      method: 'post',
      payload: {
        episodes: [{ ep: 1 }],
      },
    });

    expect(res.statusCode).toBe(401);
    expect(res.json()).toMatchObject({
      code: 'NOT_ALLOWED',
      message: expect.stringContaining('create episodes'),
    });
  });

  test('should handle non-existent subject', async () => {
    const app = await newEpisodeApp();
    const res = await app.inject({
      url: '/subjects/999999/ep',
      method: 'post',
      payload: {
        episodes: [{ ep: 1 }],
      },
    });

    expect(res.statusCode).toBe(404);
    expect(res.json()).toMatchObject({
      code: 'NOT_FOUND',
      message: expect.stringContaining('subject 999999'),
    });
  });
});
