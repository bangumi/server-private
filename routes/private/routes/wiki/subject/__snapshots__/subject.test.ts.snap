// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`edit subject  > should get current wiki info 1`] = `
Object {
  "availablePlatform": Array [
    Object {
      "id": 0,
      "text": "其他",
      "wiki_tpl": "Anime",
    },
    Object {
      "id": 1,
      "text": "TV",
      "wiki_tpl": "TVAnime",
    },
    Object {
      "id": 2,
      "text": "OVA",
      "wiki_tpl": "OVA",
    },
    Object {
      "id": 3,
      "text": "剧场版",
      "wiki_tpl": "Movie",
    },
    Object {
      "id": 5,
      "text": "WEB",
      "wiki_tpl": "TVAnime",
    },
  ],
  "id": 8,
  "infobox": "{{Infobox animanga/TVAnime
|中文名= Code Geass 反叛的鲁路修R2
|别名={
[叛逆的鲁路修R2]
[Code Geass: Hangyaku no Lelouch R2]
[叛逆的勒鲁什R2]
[叛逆的鲁鲁修R2]
[コードギアス 反逆のルルーシュR2]
[Code Geass: Lelouch of the Rebellion R2]
[叛逆的勒路什R2]
}
|话数= 25
|放送开始= 2008年4月6日
|放送星期= 
|官方网站= http://www.geass.jp/r2/
|播放电视台= 每日放送
|其他电视台= 
|播放结束= 2008年9月28日
|其他= 
|Copyright= （C）2006 SUNRISE inc./MBS
|导演= 谷口悟朗
|系列构成= 大河内一楼
|人物原案= CLAMP
|人物设定= 木村貴宏
|美术监督= 菱沼由典
|色彩设计= 岩沢れい子
|摄影监督= 大矢創太
|音响监督= 浦上靖夫、井澤基
|音乐= 中川幸太郎、黒石ひとみ
|音乐制作= AUDIO PLANNING U
|动画制作= サンライズ
}}",
  "name": "コードギアス 反逆のルルーシュR2",
  "nsfw": false,
  "platform": 1,
  "summary": "　　“东京决战”一年后，布里塔尼亚少年鲁路修在11区（原日本国）过着平凡的学生生活。但是，鲁路修与弟弟罗洛的一次出行，遇到了黑色骑士团的余党。在与少女C.C再次结成契约之后，尘封的记忆摆在了鲁路修的面前。
　　身为帝国王子的鲁路修，为了建立人人平等的世界、让妹妹娜娜丽幸福的世界，而使用GEASS，令人绝对服从的力量，带领黑色骑士团向帝国宣战。带上假面化名ZERO的他，却在一年前的“东京决战”中被好友朱雀击败，被帝国皇帝抹去了记忆。
　　现在，恢复记忆的鲁路修不仅要面对帝国的强大军事力量，更要面对虚假的弟弟罗洛、失踪的妹妹娜娜丽、不知敌友的中华联盟、内部出现分歧的黑色骑士团。面对内忧外患，鲁路修走上了GEASS之力的诅咒——孤独的王之路。 ",
  "typeID": 2,
}
`;

exports[`edit subject  > should get edit history 1`] = `
Array [
  Object {
    "commitMessage": "",
    "createdAt": 1570720134,
    "creator": Object {
      "username": "",
    },
    "type": 1,
  },
  Object {
    "commitMessage": "添加别名",
    "createdAt": 1522689610,
    "creator": Object {
      "username": "79290",
    },
    "type": 1,
  },
  Object {
    "commitMessage": "",
    "createdAt": 1430296495,
    "creator": Object {
      "username": "189074",
    },
    "type": 1,
  },
  Object {
    "commitMessage": "修正笔误",
    "createdAt": 1417061160,
    "creator": Object {
      "username": "26862",
    },
    "type": 1,
  },
  Object {
    "commitMessage": "内容扩充",
    "createdAt": 1364576814,
    "creator": Object {
      "username": "130027",
    },
    "type": 1,
  },
  Object {
    "commitMessage": "",
    "createdAt": 1357187180,
    "creator": Object {
      "username": "104510",
    },
    "type": 1,
  },
  Object {
    "commitMessage": "",
    "createdAt": 1336751754,
    "creator": Object {
      "username": "9023",
    },
    "type": 1,
  },
  Object {
    "commitMessage": "简介修正",
    "createdAt": 1325596293,
    "creator": Object {
      "username": "8766",
    },
    "type": 1,
  },
  Object {
    "commitMessage": "",
    "createdAt": 1285863983,
    "creator": Object {
      "username": "8025",
    },
    "type": 1,
  },
  Object {
    "commitMessage": "合并恢复版本",
    "createdAt": 1272719723,
    "creator": Object {
      "username": "1",
    },
    "type": 1,
  },
]
`;
